hazelcast/src/test/java/com/hazelcast/jet/SimpleTestInClusterSupport.java

```diff
@@ -129,7 +130,13 @@ public void supportAfter() {
         }
         for (HazelcastInstance instance : instances) {
             assertTrueEventually(() -> {
+                // Let's wait for all unprocessed operations (like destroying distributed object) to complete
                 assertEquals(0, getNodeEngineImpl(instance).getEventService().getEventQueueSize());
+                // Let's wait for all the jobs to stop running
+                assertEquals(0, instance().getJet().getJobs().stream()
+                        .filter(job -> job.getStatus() == JobStatus.RUNNING)
+                        .count()
+                );
```

Comment:
We call `ditchJob` for all jobs earlier, shouldn't all jobs be either COMPLETED, FAILED, CANCELLED when the execution reaches here?

Asking because a similar issue might be in other places where we call `ditchJob`

Reply:
I don't think it is a guarantee. I believe this stack:
- ```com.hazelcast.jet.impl.JobCoordinationService#terminateJob```
- ```com.hazelcast.jet.impl.MasterJobContext#requestTermination```
- ```com.hazelcast.jet.impl.MasterJobContext#handleTermination```
- ```com.hazelcast.jet.impl.MasterJobContext.ExecutionFailureCallback#cancelInvocations```
- ```com.hazelcast.jet.impl.MasterJobContext#cancelExecutionInvocations```
- ```com.hazelcast.jet.impl.MasterContext#invokeOnParticipants```
creates **asynchronous** calls to all the members. That call does:
- ```com.hazelcast.jet.impl.JobExecutionService#terminateExecution```
- ```com.hazelcast.jet.impl.JobExecutionService#terminateExecution0```
and that one completes the future. The ```getStatus()```:

```java
    @Nonnull @Override
    public final JobStatus getStatus() {
        if (isLightJob()) {
            CompletableFuture<Void> f = getFuture();
            if (!f.isDone()) {
                return JobStatus.RUNNING;
            }
            return f.isCompletedExceptionally()
                    ? JobStatus.FAILED : JobStatus.COMPLETED;
        } else {
            return getStatus0();
        }
    }
```

So as long the async call won't finish you can see ```RUNNING``` state.

Reply:
So should we move this assert to `ditchJob` so all callers of it benefit?

Reply:
Done.

Reply:
The stack trace is incomplete. The `terminateExecution0` returns a completed future typically. That futures causes a response to `TerminateExecutionOperation`, but the master doesn't wait for that responses. The master submitted also `StartExecutionOperation`, and waits for responses to that. The responses come after the `executionFuture` is completed, and that happens when all tasklets stop executing.  After the responses to all start-execution-ops arrive at master and after the terminal snapshot is done, the master calls `finalizeJob`, which calls `setFinalResult`, which completes the `jobCompletionFuture`. This future is the one to which the join operation is waiting.

Reply:
No, I can replicate it only when ```join()``` throws ```CancellationException```.

Reply:
@viliam-durina: "The master submitted also ```StartExecutionOperation```, and waits for responses to that. " - where did you find waiting for response? I see only:
```java
    private void invokeStartExecution() {
        //....
        Function<ExecutionPlan, Operation> operationCtor =
                plan -> new StartExecutionOperation(mc.jobId(), executionId, savingMetricsEnabled);
        //....
        mc.invokeOnParticipants(operationCtor, completionCallback, executionFailureCallback, false);
    }
```

but I don't see a code that would make ```mc.invokeOnParticipants``` waiting for responses.

Reply:
Waiting for responses is within `MasterContext.invokeOnParticipants()`: when it calls `invokeOnParticipant`, it gives a `remainingCount` with the number of all participants. Each participant, when it receives a response, decrements it, and when 0, calls the `completionCallback`.

Reply:
But the thread that invoked ```MasterContext.invokeOnParticipants()``` doesn't wait for that ```completionCallback```, but it continues it's work. That callback is invoked in ```future.whenCompleteAsync```, which is not executed by the thread that called ```invokeOnParticipants```.

Reply:
Yes, it's not the same thread waiting, it's async. The `jobCompletionFuture` is completed from the callback.